dashboard/app/controllers/aichat_controller.rb

```diff
@@ -14,49 +16,52 @@ def chat_completion
       return render status: :bad_request, json: {}
     end
 
+    response_body = get_response_body
+    response_body[:session_id] = log_chat_session(response_body[:messages])
+    render(status: :ok, json: response_body)
+  end
+
+  private def get_response_body
     # Check for profanity
     locale = params[:locale] || "en"
-    filter_result = ShareFiltering.find_failure(params[:newMessage], locale)
+    filter_result = ShareFiltering.find_failure(params[:newMessage][:chatMessageText], locale)
     if filter_result&.type == ShareFiltering::FailureType::PROFANITY
-      new_messages = [
-        {
-          role: 'user',
-          content: params[:newMessage],
-          status: SharedConstants::AI_INTERACTION_STATUS[:PROFANITY_VIOLATION]
-        }
+      messages = [
+        get_user_message(SharedConstants::AI_INTERACTION_STATUS[:PROFANITY_VIOLATION])
       ]
-      session_id = log_chat_session(new_messages)
-      return render(
-        status: :ok,
-        json: {
-          status: SharedConstants::AICHAT_ERROR_TYPE[:PROFANITY_USER],
-          flagged_content: filter_result.content,
-          session_id: session_id
-        }
-      )
+
+      return {
+        messages: messages,
+        flagged_content: filter_result.content,
+      }
+    end
+
+    messages_for_model = params.to_unsafe_h[:storedMessages].filter do |message|
+      message[:status] == SharedConstants::AI_INTERACTION_STATUS[:OK] &&
+        ROLES_FOR_MODEL.include?(message[:role])
     end
 
     # Use to_unsafe_h here to allow testing this function.
     # Safe params are primarily targeted at preventing "mass assignment vulnerability"
     # which isn't relevant here.
     input = AichatSagemakerHelper.format_inputs_for_sagemaker_request(
       params.to_unsafe_h[:aichatModelCustomizations],
-      params.to_unsafe_h[:storedMessages].filter {|message| message[:status] == SharedConstants::AI_INTERACTION_STATUS[:OK]},
+      messages_for_model,
       params.to_unsafe_h[:newMessage]
     )
     sagemaker_response = AichatSagemakerHelper.request_sagemaker_chat_completion(input, params[:aichatModelCustomizations][:selectedModelId])
     latest_assistant_response = AichatSagemakerHelper.get_sagemaker_assistant_response(sagemaker_response)
 
     filter_result = ShareFiltering.find_failure(latest_assistant_response, locale)
     if filter_result&.type == ShareFiltering::FailureType::PROFANITY
-      new_messages = [
+      messages = [
+        get_user_message(SharedConstants::AI_INTERACTION_STATUS[:ERROR]),
         {
-          role: 'user',
-          content: params[:newMessage],
-          status: SharedConstants::AI_INTERACTION_STATUS[:ERROR]
+          role: "assistant",
+          status: SharedConstants::AI_INTERACTION_STATUS[:ERROR],
+          chatMessageText: '[redacted - model generated profanity]',
```

Comment:
Nit: Wonder if we should defined 'redacted' for users.

Reply:
Sorry, not sure I'm following...what do you mean "define"?

Reply:
Also maybe worth noting that this message doesn't actually get displayed to users -- we just display a generic "error getting response" message.

Reply:
Oh - gotcha. Sounds good!